docs/design/vm_selective_velero_restore_mechanism.md

```diff
@@ -0,0 +1,205 @@
+- Extends Velero to support selective backup and restore of KubeVirt data-only resources.
+- Handles the data management logic transparently.
+
+**VMFR Controller:**
+- Triggers restores by creating Velero `Restore` objects.
+- Provides the necessary configuration expected by the plugin to selectively restore objects associated with a given VM.
+
+### 4.1. Backup Phase: Kubevirt Velero Plugin Enhancements
+
+The existing KubeVirt Velero plugin will be enhanced to embed crucial metadata into backup objects. This metadata exists solely within the Velero backup and serves as the primary mechanism for scoping restore operations. It is never persisted to live cluster resources, and during a restore, the plugin uses it to guide the process while ensuring that the restored objects match their original state, without retaining any temporary backup-only data.
+
+### 4.1.1. Metadata Strategy (Labels and Annotations)
+
+The KubeVirt Velero plugin will extend backup and restore actions for PVs and PVCs to manage additional labels and annotations used for selective restore. 
+
+**Backup Action:**
+- Add labels to PVs and PVCs (e.g., `kubevirt.io/pvc-name`, `kubevirt.io/pv-name`) to record object identity.
+- Before adding a label:
+  - Preserve any existing label with the same key but different value.
+  - If an existing label had a different value, record its original value in an annotation so it can be restored later.
+- Ensure that object labels and annotations prior to the backup remain fully restorable.
+
+**Restore Action:**
+- Remove the temporary labels added during backup.  
+- If the object had original labels that were overwritten, restore them to their original values.  
+- Do not remove or modify labels that were identical to the backup-added labels unless they were added by the plugin itself.
+- Remove stored information.
+- Guarantee that after restore, the PV/PVC objects match exactly their state prior to the backup.
+
+**Notes:**
+- The logic is implemented in:
+  - `pkg/plugin/pv_backup_item_action.go` / `_test.go`
+  - `pkg/plugin/pv_restore_item_action.go` / `_test.go`
+  - `pkg/plugin/pvc_backup_item_action.go` / `_test.go`
+  - `pkg/plugin/pvc_restore_item_action.go` / `_test.go`
+
+
+No VM-specific labels are applied to shared resources. The VMFR Controller is responsible for mapping VirtualMachines to their associated resources based on this backup metadata, ensuring correct selective restore even when multiple VMs share the same PV/PVC.
+
+### 4.1.2. Example 1: Non-conflicting labels
+
+Original object (pseudo code)
+
+```yaml
+Namespace: virtualmachines
+VirtualMachine: production-vm
+    PVC name: data-pvc
+      Labels: {}
+      Annotations: {}
+    PV name: pvc-1234
+      Labels: {}
+      Annotations: {}
+```
+
+**Backup operation changes:**
+
+```yaml
+Namespace: virtualmachines
+VirtualMachine: production-vm
+    PVC name: data-pvc
+      Labels:
+        - kubevirt.io/pvc-name: data-pvc
```

Comment:
I do wonder if there should be embedded somewhere namespace as well? Trying to think of a conflict where backup may have backed up multiple objects such as pvc's from multiple namespaces and those had exactly same name.

During restore operation we would not know which exact object to include.

This however will require similar logic with additional label and annotation to store the relevant information:
```
kubevirt.io/pvc-namespace: virtualmachines
```

It may be not required at all, thinking here of edge cases.